qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -81,32 +122,110 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
 
         Returns:
             Output dag with UnitaryGates synthesized to target basis.
+
+        Raises:
+            TranspilerError:
+                1. pulse_optimize is True but pulse optimal decomposition is not known
+                   for requested basis.
+                2. pulse_optimize is True and natural_direction is True but a preferred
+                   gate direction can't be determined from the coupling map or the
+                   relative gate lengths.
         """
+
         euler_basis = _choose_euler_basis(self._basis_gates)
         kak_gate = _choose_kak_gate(self._basis_gates)
-
         decomposer1q, decomposer2q = None, None
         if euler_basis is not None:
             decomposer1q = one_qubit_decompose.OneQubitEulerDecomposer(euler_basis)
         if kak_gate is not None:
-            decomposer2q = TwoQubitBasisDecomposer(kak_gate, euler_basis=euler_basis)
-
-        for node in dag.named_nodes("unitary"):
+            decomposer2q = TwoQubitBasisDecomposer(
+                kak_gate, euler_basis=euler_basis, pulse_optimize=self._pulse_optimize
+            )
 
+        for node in dag.named_nodes(*self._synth_gates):
+            if self._basis_gates and node.name in self._basis_gates:
+                continue
             synth_dag = None
+            wires = None
             if len(node.qargs) == 1:
                 if decomposer1q is None:
                     continue
                 synth_dag = circuit_to_dag(decomposer1q._decompose(node.op.to_matrix()))
             elif len(node.qargs) == 2:
                 if decomposer2q is None:
                     continue
-                synth_dag = circuit_to_dag(
-                    decomposer2q(node.op.to_matrix(), basis_fidelity=self._approximation_degree)
-                )
+                synth_dag, wires = self._synth_natural_direction(node, dag, decomposer2q)
             else:
                 synth_dag = circuit_to_dag(isometry.Isometry(node.op.to_matrix(), 0, 0).definition)
 
-            dag.substitute_node_with_dag(node, synth_dag)
+            dag.substitute_node_with_dag(node, synth_dag, wires=wires)
 
         return dag
+
+    def _synth_natural_direction(self, node, dag, decomposer2q):
+        layout = self.property_set["layout"]
+        natural_direction = None
+        synth_direction = None
+        physical_gate_fidelity = None
+        wires = None
+        if self._natural_direction in {None, True} and layout and self._coupling_map:
+            neighbors0 = self._coupling_map.neighbors(dag.qubits.index(node.qargs[0]))
```

Comment:
Instead of calling `.index()` on every iteration it'd be better to build a dict mapping the bit to index once at the top of the function with something like: `{bit: index for index, bit in enumerate(dag.qubits)}` then you can replace all the `.index()` calls with a dict lookup